selfdrive/ui/replay/replay.cc

```diff
@@ -25,37 +72,280 @@ void Replay::parseResponse(QString response){
   camera_paths = doc["cameras"].toArray();
   log_paths = doc["logs"].toArray();
 
-  // add first segment
-  addSegment(0);
+  seek_queue.enqueue({false, seek});
 }
 
 void Replay::addSegment(int i){
   if (lrs.find(i) != lrs.end()) {
     return;
   }
 
-  QThread* thread = new QThread;
+  QThread* lr_thread = new QThread;
 
   QString log_fn = this->log_paths.at(i).toString();
-  lrs.insert(i, new LogReader(log_fn, &events, &events_lock, &unlogger->eidx));
+  lrs.insert(i, new LogReader(log_fn, &events, &events_lock, &eidx));
 
-  lrs[i]->moveToThread(thread);
-  QObject::connect(thread, SIGNAL (started()), lrs[i], SLOT (process()));
-  thread->start();
+  lrs[i]->moveToThread(lr_thread);
+  QObject::connect(lr_thread, SIGNAL (started()), lrs[i], SLOT (process()));
+  lr_thread->start();
 
   QString camera_fn = this->camera_paths.at(i).toString();
   frs.insert(i, new FrameReader(qPrintable(camera_fn)));
 }
 
+void Replay::trimSegment(int seg_num){
+  lrs.remove(seg_num);
+  frs.remove(seg_num);
+
+  auto eit = events.begin();
+  while(eit != events.end()){
+    if(std::abs(eit.key()/1e9 - getCurrentTime()/1e9) > window_padding*60.0){
+      eit = events.erase(eit);
+      continue;
+    }
+    eit++;
+  }
+}
+
 void Replay::stream(SubMaster *sm){
-  QThread* thread = new QThread;
-  unlogger->moveToThread(thread);
+  thread = new QThread;
+  this->moveToThread(thread);
   QObject::connect(thread, &QThread::started, [=](){
-    unlogger->process(sm);
+    process(sm);
   });
   thread->start();
 
-  QObject::connect(unlogger, &Unlogger::loadSegment, [=](){
-    addSegment(++current_segment);
+  seek_thread = new QThread;
+  QObject::connect(seek_thread, &QThread::started, [=](){
+    seekThread();
+  });
+  seek_thread->start();
+
+  queue_thread = new QThread;
+  QObject::connect(queue_thread, &QThread::started, [=](){
+    seekRequestThread();
   });
+  queue_thread->start();
+}
+
+void Replay::seekTime(int seek_){
+  // TODO: see if eidx also needs to be cleared
+  if(!seeking){
+    if(seek >= 0){
+      setSeekRequest(seek_*1e9);
+    }
+
+    if(seek_/60 != current_segment) {
+      int last_segment = current_segment;
+      current_segment = seek_/60;
+
+      for(int i = 0 ; i < 2*window_padding + 1 ; i++) {
```

Comment:
parameterize the window

Reply:
What do you mean? This is already parameterized